### ISSUE TITLE
VNC does  not work with Docker for Mac.

### ISSUE THREAD

**Issue by JoyceBabu** [aerokube/images#374]
I have the following `docker-compose.yml`, which works fine on linux. But the same is not working with Docker for Mac.

### docker-compose.yml
```
version: "3.5"
services:
  selenoid:
    container_name: selenoid
    image: "aerokube/selenoid:latest-release"
    networks:
      - selenoid
    ports:
      - "4444:4444"
    volumes:
      - "$PWD:/etc/selenoid"
      - "/var/run/docker.sock:/var/run/docker.sock"
      - "$PWD/output/video:/opt/selenoid/video"
      - "$PWD/output/logs:/opt/selenoid/logs"
      - "./browsers.json:/etc/selenoid/browsers.json"
    command: [
      "-capture-driver-logs", "-save-all-logs",
      "-conf", "/etc/selenoid/browsers.json",
      "-video-output-dir", "/opt/selenoid/video",
      "-log-output-dir", "/opt/selenoid/logs",
      "-video-recorder-image", "selenoid/video-recorder:latest-release",
      "-limit", "20",
      "-listen", "0.0.0.0:4444",
      "-container-network", "selenoid"
    ]
    environment:
      - OVERRIDE_VIDEO_OUTPUT_DIR=$PWD/output/video

  selenoid-ui:
    container_name: selenoid-ui
    image: "aerokube/selenoid-ui"
    networks:
      - selenoid
    links:
      - selenoid
    ports:
      - "3030:8080"
    command: ["--selenoid-uri", "http://selenoid:4444"]

networks:
    selenoid:
        name: selenoid
```

### browsers.json
```
{
  "chrome": {
    "default": "89.0",
    "versions": {
      "89.0": {
        "image": "selenoid/vnc:chrome_89.0",
        "port": "4444",
        "path": "/",
        "tmpfs": {
          "/tmp": "size=128m"
        }
      }
    }
  }
}
```

On Docker for Mac, the log is

```
selenoid       | 2021/04/11 08:14:16 [-] [NEW_REQUEST] [unknown] [172.22.0.1]
selenoid       | 2021/04/11 08:14:16 [-] [NEW_REQUEST_ACCEPTED] [unknown] [172.22.0.1]
selenoid       | 2021/04/11 08:14:16 [11] [LOCATING_SERVICE] [chrome] [84.0]
selenoid       | 2021/04/11 08:14:16 [11] [USING_DOCKER] [chrome] [84.0]
selenoid       | 2021/04/11 08:14:16 [11] [CREATING_CONTAINER] [selenoid/vnc:chrome_84.0]
selenoid       | 2021/04/11 08:14:16 [11] [STARTING_CONTAINER] [selenoid/vnc:chrome_84.0] [ad74d4f00ecb09e15f02a174400569d1c6ff2349e40b6ddca2cdb3cbcde88b6a]
selenoid       | 2021/04/11 08:14:18 [11] [CONTAINER_STARTED] [selenoid/vnc:chrome_84.0] [ad74d4f00ecb09e15f02a174400569d1c6ff2349e40b6ddca2cdb3cbcde88b6a] [2.09s]
selenoid       | 2021/04/11 08:14:48 [11] [REMOVING_CONTAINER] [ad74d4f00ecb09e15f02a174400569d1c6ff2349e40b6ddca2cdb3cbcde88b6a]
selenoid       | 2021/04/11 08:14:48 [11] [CONTAINER_REMOVED] [ad74d4f00ecb09e15f02a174400569d1c6ff2349e40b6ddca2cdb3cbcde88b6a]
selenoid       | 2021/04/11 08:14:48 [11] [SERVICE_STARTUP_FAILED] [wait: http://172.22.0.4:4444/ does not respond in 30s]
```

Whereas on Linux, the log is

```
selenoid       | 2021/04/11 08:11:23 [-] [NEW_REQUEST] [unknown] [192.168.0.4]
selenoid       | 2021/04/11 08:11:23 [-] [NEW_REQUEST_ACCEPTED] [unknown] [192.168.0.4]
selenoid       | 2021/04/11 08:11:23 [3] [LOCATING_SERVICE] [chrome] [84.0]
selenoid       | 2021/04/11 08:11:23 [3] [USING_DOCKER] [chrome] [84.0]
selenoid       | 2021/04/11 08:11:23 [3] [CREATING_CONTAINER] [selenoid/vnc:chrome_84.0]
selenoid       | 2021/04/11 08:11:23 [3] [STARTING_CONTAINER] [selenoid/vnc:chrome_84.0] [04a958c98a9e8dbcdf777d54de219c923abde58d82b8460674deda44ff76b88b]
selenoid       | 2021/04/11 08:11:24 [3] [CONTAINER_STARTED] [selenoid/vnc:chrome_84.0] [04a958c98a9e8dbcdf777d54de219c923abde58d82b8460674deda44ff76b88b] [0.99s]
selenoid       | 2021/04/11 08:11:25 [3] [SERVICE_STARTED] [selenoid/vnc:chrome_84.0] [04a958c98a9e8dbcdf777d54de219c923abde58d82b8460674deda44ff76b88b] [0.76s]
selenoid       | 2021/04/11 08:11:25 [3] [PROXY_TO] [04a958c98a9e8dbcdf777d54de219c923abde58d82b8460674deda44ff76b88b] [http://172.25.0.4:4444/]
selenoid       | 2021/04/11 08:11:25 [3] [SESSION_ATTEMPTED] [http://172.25.0.4:4444/] [1]
selenoid       | 2021/04/11 08:11:26 [3] [SESSION_CREATED] [178c3211afef2445809cd68f776e4824] [1] [2.53s]
selenoid-ui    | 2021/04/11 08:11:30 [WS_PROXY] [/ws/logs/178c3211afef2445809cd68f776e4824] [ws://selenoid:4444/logs/178c3211afef2445809cd68f776e4824]
selenoid       | 2021/04/11 08:11:30 [7] [CONTAINER_LOGS] [04a958c98a9e8dbcdf777d54de219c923abde58d82b8460674deda44ff76b88b]
selenoid-ui    | 2021/04/11 08:11:30 [WS_PROXY] [/ws/vnc/178c3211afef2445809cd68f776e4824] [ws://selenoid:4444/vnc/178c3211afef2445809cd68f776e4824]
selenoid       | 2021/04/11 08:11:30 [8] [VNC_ENABLED] [178c3211afef2445809cd68f776e4824]
```

When I tried to run the image `selenoid/vnc:chrome_84.0` directly on Docker for Mac, I got the following error

```
docker run -ti -e TZ=UTC -e ENABLE_VNC=true -e ENABLE_VIDEO=false -e DBUS_SESSION_BUS_ADDRESS=/dev/null selenoid/vnc:chrome_84.0
Waiting X server...
2021/04/11 08:18:31 [INIT] [Listening on :7070]
Waiting X server...
Logging to: /home/selenium/fluxbox.log
Waiting X server...
Waiting X server...
Waiting X server...

 --- x11vnc loop: 1 ---

Starting ChromeDriver 84.0.4147.30 (48b3e868b4cc0aa7e8149519690b6f6949e110a8-refs/branch-heads/4147@{#310}) on port 4444
Search: XIO (1/2 matches. case-sensitive)
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.ChromeDriver was started successfully.
 --- x11vnc loop: waiting for: 72

PORT=5900

^CXIO:  fatal IO error 11 (Resource temporarily unavailable) on X server ":99"
      after 4859 requests (4859 known processed) with 0 events remaining.
```

**Comment by vania-pooh** [comment]
@JoyceBabu do you have sufficient memory available to Docker VM on Mac?

**Comment by JoyceBabu** [comment]
Docker for mac is configured with 6 GB RAM and 1.5GB Swap.

**Comment by JoyceBabu** [comment]
I checked the log output from the browser container, when it was running and found an error related to `/dev/null`.

```
/entrypoint.sh: line 50: cannot redirect standard input from /dev/null: Permission denied
/entrypoint.sh: line 56: cannot redirect standard input from /dev/null: Permission denied
/entrypoint.sh: line 53: cannot redirect standard input from /dev/null: Permission denied
/entrypoint.sh: line 59: /dev/null: Permission denied
/entrypoint.sh: line 61: cannot redirect standard input from /dev/null: Permission denied
/entrypoint.sh: line 61: /dev/null: Permission denied
/entrypoint.sh: line 66: /dev/null: Permission denied
Waiting X server...
2021/04/11 18:38:27 [INIT] [Listening on :7070]
Waiting X server...
/entrypoint.sh: line 66: /dev/null: Permission denied
/entrypoint.sh: line 66: /dev/null: Permission denied
Waiting X server...
/entrypoint.sh: line 66: /dev/null: Permission denied
Waiting X server...
/entrypoint.sh: line 66: /dev/null: Permission denied
Waiting X server...
/entrypoint.sh: line 66: /dev/null: Permission denied
Waiting X server...
/entrypoint.sh: line 66: /dev/null: Permission denied
Waiting X server...
```

So, I checked the permission on `/dev/null` by starting the `selenoid/vnc:chrome_84.0` manually, and found that it has correct permission.

```
docker run --entrypoint bash selenoid/vnc:chrome_84.0 -c 'ls -l /dev/null'
crw-rw-rw- 1 root root 1, 3 Apr 11 18:50 /dev/null
```

But the permission is incorrect on the container started from selenoid ui.

```
docker ps | fgrep vnc | head -1 | awk '{print $1}'| xargs -I{} docker exec {} ls -l /dev/null
crw-rw---- 1 root root 1, 3 Apr 11 18:43 /dev/null
```

**Comment by JoyceBabu** [comment]
I created a custom docker image and modified `entrypoint.sh` to add a `sleep 5` statement the beginning. And during this this, I tried fixing the permission of  `/dev/null` and this solved the startup errors that I reported earlier.

But chromedriver was failing with error
> Trace/breakpoint trap

A quick strace revealed the following error
> openat(AT_FDCWD, "/dev/urandom", O_RDONLY|O_CLOEXEC) = -1 EACCES (Permission denied)

So I fixed the permissions on both `/dev/null` and `/dev/urandom` and now I am able to open and access the browser interface from selenoid ui.

What is causing the file permission modificiation?
